### PR TITLE
Expose apply_ufunc as public API and add documentation

### LINKED ISSUE #770: Internal refactor: create a generic function for applying ufuncs-like functions to xarray objects
It would be awesome to have a generic function for making functions that act like NumPy's generalized universal functions "xarray aware".

What would `xarray.apply_ufunc(func, objs, join='inner', agg_dims=None, drop_dims=None, kwargs=None)` do?
1. If one or more of the provided objects are `Dataset` or `GroupBy` instances, dispatch to specialized loops that call the remainder of `apply_ufunc` repeatedly.
2. `align` all objects along shared labels using the indicated `join` (for some operations, e.g., `where`, a left join is appropriate rather than an inner join).
3. `broadcast` all objects against each other to expand dimensionality along all dimensions except (optionally) those listed in `agg_dims`/`drop_dims`. `drop_dims` should be moved to the end, for consistency with gufunc signatures.
4. Transform `agg_dims` (if provided) into an `axis` argument using `get_axis_num` and insert it into `kwargs`.
5. Apply `func` to the `data` argument of each array to calculate the result using the provided `kwargs`. The result is expected to have all the same dimensions in the provided arrays, except any listed in the `dims` and `drop_dims` arguments.
6. `merge` all coordinate data together (i.e., with an n-ary version of the `Coordinate.merge` method) and add these to the result array.

If any of `args` are not xarray objects (e.g., they're NumPy or dask arrays), they should be skipped in operations that don't apply to them. `xarray.Variable` don't align or have coordinates, for example.

A concrete example of similar functionality in dask.array is [`atop`](https://github.com/dask/dask/blob/0.8.0/dask/array/core.py#L1572). The most similar thing to this that we currently have in xarray are the `_unary_op` and `_binary_op` staticmethods (e.g., [on DataArray](https://github.com/pydata/xarray/blob/v0.7.1/xarray/core/dataarray.py#L1180)), but these only handle one or two arguments, don't handle aggregated dimensions and most importantly, are difficult to apply to new operations.

Here are a few concrete examples of how this could work:

``` python
def average(array, weights, dim=None):
    # still needs a bit of work to make a NaN and dask.array safe version
    # version of np.average 
    return apply_ufunc(np.average, [array, weights], agg_dims=dim)

def where(cond, first, second=None):
    if second is None:
        # need to write where2, a function that looks at first.dtype
        # to infer the appropriate NA sentinel value
        return apply_ufunc(ops.where2, [cond, first])
    else:
        return apply_ufunc(ops.where, [cond, first, second])

def dot(self, other, dim=None):
    if dim is None:
        dim = set(self.dims) ^ set(other.dims)
    return apply_ufunc(ops.tensordot, [self, other], agg_dims=dim)
```

### PR DESCRIPTION
**Pull request by shoyer**
`apply_ufunc()` does not meet all our needs for wrapping unlabeled array routines with xarray (see #1618 for a proposal for `apply_raw()`), but it should be useful for many advanced users and isn't doing much in its current state as non-public API. So I'd like to make it public for the xarray 0.10 release.

 - [x] Closes #770
 - [x] Passes ``git diff upstream/master | flake8 --diff``
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API
 - [x] Turn the example from #1517 into a separate docs page of some sort, and link to it from `computation.rst` and `dask.rst`.

@MaximilianR @rabernat @jhamman Review from any of you would be appreciated here!

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -75,6 +75,11 @@ Backward Incompatible Changes
 Enhancements
 ~~~~~~~~~~~~
 
+- New helper function :py:func:`~xarray.apply_ufunc` for wrapping functions
+  written to work on NumPy arrays to support labels on xarray objects.
+  ``apply_ufunc`` also support automatic parallelization for many functions
+  with dask. See :ref:`computation.wrapping-custom` for details.
+
```
Comment (jhamman) [inline doc/whats-new.rst]: Don't forget: 
```
(:issue:`XXXX`). By `Stephan Hoyer <https://github.com/shoyer>`_. 
```

File: doc/computation.rst [hunk 2]
```diff
@@ -329,5 +335,93 @@ Arithmetic between two datasets matches data variables of the same name:
...
+Because ``apply_ufunc`` follows a standard convention for ufuncs, it plays
+nicely with tools for building vectorized functions, like
+:func:`numpy.broadcast_arrays` and :func:`numpy.vectorize`. For high performance
+needs, consider using Numba's `vectorize and guvectorize <http://numba.pydata.org/numba-doc/dev/user/vectorize.html>`_.
+
+In addition to wrapping functions, ``apply_ufunc`` can automatically parallelize
+many functions when using dask by setting ``dask='parallelized'``. This is
+illustrated in a separate example.
+
+.. TODO: add link!
```
Comment (jhamman) [inline doc/computation.rst]: I'm thinking the recipes would be a good place for this: http://xarray.pydata.org/en/stable/auto_gallery/index.html

Comment (shoyer) [reply doc/computation.rst]: Yes, if I can come up with a figure!

File: doc/computation.rst [hunk 3]
```diff
@@ -303,15 +303,21 @@ Datasets support most of the same methods found on data arrays:
     ds.mean(dim='x')
     abs(ds)
 
-Unfortunately, a limitation of the current version of numpy means that we
-cannot override ufuncs for datasets, because datasets cannot be written as
-a single array [1]_. :py:meth:`~xarray.Dataset.apply` works around this
+Unfortunately, we current do not support NUmPy ufuncs for datasets [1]_.
```
Comment (benbovy) [inline doc/computation.rst]: currently

File: doc/computation.rst [hunk 4]
```diff
@@ -329,5 +335,93 @@ Arithmetic between two datasets matches data variables of the same name:
...
+functions/methods are written using ``apply_ufunc``.
+
+Simple functions that act independently on each value should work without
+any additional arguments:
+
+.. ipython:: python
+
+    squared_error = lambda x, y: (x - y) ** 2
+    arr1 = xr.DataArray([0, 1, 2, 3], dims='x')
+    xr.apply_func(squared_error, arr1, 1)
```
Comment (benbovy) [inline doc/computation.rst]: `xr.apply_ufunc`

**Comment by shoyer** [comment]
@benbovy @jhamman thanks for giving this a read. The docs actually build now, and I added a section on "Automatic parallelization" to the dask docs.

File: doc/dask.rst [hunk 5]
```diff
@@ -213,6 +213,93 @@ loaded into dask or not:
...
+-------------------------
+
+Almost all of xarray's built-in operations work on dask arrays. If you want to
+use a function that isn't wrapped by xarray, one option is to extract dask
+arrays from xarray objects (``.data``) and use dask directly.
+
+Another option is to use xarray's :py:func:`~xarray.apply_ufunc`, which can
+automatically parallelize functions written for processing NumPy arrays when
+applied to dask arrays. It works similarly to :py:func:`dask.array.map_blocks`
+and :py:func:`dask.array.atop`, but without requiring an immediate layer of
```
Comment (jhamman) [inline doc/dask.rst]: did you mean intermediate?

Comment (shoyer) [reply doc/dask.rst]: yes, indeed

File: doc/computation.rst [hunk 6]
```diff
@@ -329,5 +335,91 @@ Arithmetic between two datasets matches data variables of the same name:
...
+       (:issue:`1617`).
+
+.. _comput.wrapping-custom:
+
+Wrapping custom computation
+===========================
+
+It doesn't always make sense to do computation directly with xarray objects:
+
+  - When working with small arrays (less than ~1e7 elements), applying an
```
Comment (MaximilianR) [inline doc/computation.rst]: Is the point on speed is a distraction? If an array is that small, the _absolute_ difference in speed is still very small, so it really only makes a difference if you're doing those operations in a loop.

File: xarray/core/computation.py [hunk 7]
```diff
@@ -724,7 +727,7 @@ def apply_ufunc(func, *args, **kwargs):
     ``apply_ufunc`` to write functions to (very nearly) replicate existing
     xarray functionality:
 
-    Calculate the vector magnitude of two arguments:
+    Calculate the vector magnitude of two arguments::
```
Comment (MaximilianR) [inline xarray/core/computation.py]: ?

File: doc/computation.rst (hunk 6 above)
Comment (shoyer) [reply doc/computation.rst]: Agreed. I had "in the inner loop" in mind when I wrote this, but I see now that that never made it into the text. Let me know if this latest update seems more reasonable to you, or if I'm still pushing too hard on performance considerations.

**Comment by shoyer** [comment]
Any additional comments? If not, I'll merge this in about 24 hours.

**Pull request merged**